Core/src/IfcGeometryConverter/SplineConverter.h

```diff
@@ -147,12 +147,149 @@ namespace OpenInfraPlatform
-						const std::vector<std::vector<carve::geom::vector<3>>>& controlPoints,
+						const EXPRESSReference<typename IfcEntityTypesT::IfcBSplineSurface>& splineSurface,
+						const carve::math::Matrix& pos, // AT THE MOMENT, NO IDEA WHAT THIS IS
+						//const std::vector<std::vector<carve::geom::vector<3>>>& controlPoints,
 						std::shared_ptr<carve::input::PolylineSetData>& polylineData) const throw(...)
 					{
-						// not implemented, code in commented out part at the end of the file
-						throw oip::UnhandledException(splineSurface);
+
+						if (splineSurface.isOfType<typename IfcEntityTypesT::IfcBSplineSurfaceWithKnots>())
+						{
+							const EXPRESSReference<typename IfcEntityTypesT::IfcBSplineSurfaceWithKnots>& bsplineSurfaceWithKnots =
+								splineSurface.as<typename IfcEntityTypesT::IfcBSplineSurfaceWithKnots>();
+
+							// obtain degree of both b-spline curves
+							const int degreeU = bsplineSurfaceWithKnots->UDegree;
+							const int orderU = degreeU + 1;
+							const int degreeV = bsplineSurfaceWithKnots->VDegree;
+							const int orderV = degreeV + 1;
+							// obtain number of control points
+							//const int numControlPointsU = controlPoints.size();
+							//const int numControlPointsV = controlPoints[0].size();
+
+							// load array of control points from ifc-entity
+							// simplified, it's in the type EXPRESSContainer<EXPRESSContainer<IfcCartesianPoint>>
+							const auto controlPointList = bsplineSurfaceWithKnots->ControlPointsList;
+							// temporary control point container of type std::vector<std::vector<..>> which is needed for converter function
+							std::vector<std::vector<EXPRESSReference<typename IfcEntityTypesT::IfcCartesianPoint>>> controlPointListVector;
+							controlPointListVector.resize(controlPointList.size());
+							// convert from EXPRESSContainer<EXPRESSContainer<..>> to std::vector<std::vector<..>>
+							// loop over u-indeces (=K1) of control points
+							for (size_t i = 0; i < controlPointList.size(); i++)
+							{
+								controlPointListVector[i].resize(controlPointList[i].size());
+								// loop over v-indeces (=K2) of control points
+								for (size_t j = 0; j < controlPointList[i].size(); j++)
+								{
+									controlPointListVector[i][j] = controlPointList[i][j];
+								}
+							}
+
+							// temporary instances to curve and face converter, potential relocation will be considered in further refactoring
+							std::shared_ptr<CurveConverterT<IfcEntityTypesT>> curveConverter =
+								std::make_shared<CurveConverterT<IfcEntityTypesT>>(CurveConverterT<IfcEntityTypesT>(GeomSettings(), UnitConvert(), placementConverter));
+							FaceConverterT<IfcEntityTypesT> faceConverter(GeomSettings(), UnitConvert(), placementConverter, curveConverter);
+
+							// obtain control points
+							const std::vector<std::vector<carve::geom::vector<3>>> controlPoints = faceConverter.convertIfcCartesianPoint2DVector(controlPointListVector);

```

Comment:
I see that the transformation from `EXPRESSContainer` to `std::vector` is necessary because of this function call. I suggest changing the function to take `EXPRESSContainer`.

Reply:
Done in commit 9b4c92d.